src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/SmallFrozenDictionary.cs

```diff
@@ -24,8 +25,8 @@ internal SmallFrozenDictionary(Dictionary<TKey, TValue> source) : base(source.Co
         {
             Debug.Assert(source.Count != 0);
 
-            _keys = source.Keys.ToArray(source.Count);
-            _values = source.Values.ToArray(source.Count);
+            _keys = source.Keys.ToArray();
+            _values = source.Values.ToArray();
```

Comment:
Isn't this heavily based on the fact that MaxItemsInSmallFrozenCollection is currently 4, which happens to be exactly the default initial array allocation size in ToArray?  If MaxItemsInSmallFrozenCollection gets changed to 8, for example, will this change still be a win for inputs of length 8?

Reply:
The custom `ToArray` implementation is slower, because it uses `CopyTo` only for collections that implement `IList<T>`, while both `KeyCollection` and `ValueCollection` don't. They do implement `ICollection<T>` which is used by LINQ.

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.Minimal.cs#L158-L164

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.Minimal.cs#L106

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/System.Collections/ref/System.Collections.cs#L504

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/System.Collections/ref/System.Collections.cs#L532

Reply:
> because it uses CopyTo only for collections that implement IList<T>

We should revisit this.
https://github.com/dotnet/runtime/blob/e5ddc6c9b7e481d36d6350960868188d7b275ef7/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.Minimal.cs#L93-L96
That logic, while well-meaning, isn't worth making the implementations slower.  If code wants to get at the underlying array of an ImmutableArray, there are easier ways, including a now fully supported public API.

Reply:
I agree, but I don't believe it's in the scope of this PR, so I've created a dedicated issue for it: https://github.com/dotnet/runtime/issues/88181